doc/dev/debugging.rst

```diff
@@ -0,0 +1,197 @@
+
+.. code-block:: C
+
+    #include <complex.h>
+    #include <stddef.h>
+    #include <stdint.h>
+
+    typedef int64_t int64;
+    typedef double float64;
+    typedef float float32;
+    typedef double complex complex128;
+    typedef float complex complex64;
+
+
+    struct memref_float64x0_t
+    {
+        float64* allocated;
+        float64* aligned;
+        size_t offset;
+
+    };
+    struct result_t {
+        struct memref_float64x0_t f_0;
+    };
+
+
+    extern void setup(int, char**);
+    extern void _catalyst_ciface_jit_identity(struct result_t*, struct memref_float64x0_t*);
+    extern void teardown();
+
+    int
+    main(int argc, char** argv)
+    {
+
+        struct result_t result_val;
+        float64 buff_0 = 1.0;
+        struct memref_float64x0_t arg_0 = { &buff_0, &buff_0, 0 };
+
+
+        setup(1, &argv[0]);
+        _catalyst_ciface_jit_identity(&result_val, &arg_0);
+        teardown();
+    }
+
+The user can now compile and link this program and run without ``python``.
+
+Pass Pipelines
+==============
+
+A lot of the compilation steps are broken into pass pipelines.
+``PassPipeline`` is a class that specifies which binary and which flags are used for compilation.
+``PassPipeline`` can be implemented by a user if the user wants to insert new passes in between defaul ``PassPipeline``'s.
+We won't get into too much detail here, but sometimes it is useful to look at the output of a specific ``PassPipeline``.
+To do so, simply use the ``get_output_of`` method available in ``QJIT``.
+For example, if one wishes to inspect the output of the ``BufferizationPass``, simply run the following command.
+
+.. code-block:: python
+
+    circuit.get_output_of("BufferizationPass")
+
+
+Compilation Steps
+=================
+
+The compilation process of a QJITed quantum function moves through various stages of the compilation pipeline including:
+
+- **Quantum Tape**: The quantum record of variational quantum programs in a single ``qml.QNode``
+- **JAXPR**: The graph data structure maintained by `JAX <https://github.com/google/jax>`_ of the classical part for transformations
+- **MLIR**: A novel compiler framework and intermediate representation
+- **HLO (XLA) + Quantum Dialect**: Lowering to `HLO <https://github.com/tensorflow/mlir-hlo>`_ is the first stage inside MLIR after leaving JAXPR
+- **Builtin + Quantum Dialects**: HLO has been converted to a variety of classical dialects in MLIR
+- **Bufferized MLIR**: All tensors have been `converted <https://mlir.llvm.org/docs/Bufferization>`_ to memory buffer allocations at this step
+- **LLVM Dialect**: Lower the code to `LLVM Dialect <https://mlir.llvm.org/docs/Dialects/LLVM/>`_ to target LLVMIR by providing a one-to-one mapping in MLIR
+- **QIR**: A `specification <https://learn.microsoft.com/en-us/azure/quantum/concepts-qir>`_ for quantum programs in LLVMIR
+
+To ensure that you have access to all the stages, the ``keep_intermediate=True`` flag must be specified in ``qjit``.
```

Comment:
```suggestion
To ensure that you have access to all the stages, the ``keep_intermediate=True`` flag must be specified in the ``qjit`` decorator.
```

Reply:
I thought this was no longer a requirement, since the temp files are kept alive as long as the QJIT is alive?

Reply:
Both statements can considered correct, the question is "when does the user have access to all the stages?". The temp files are alive as long as the QJIT object has not being garbage collected in Python. However, if you want to see the files after the QJIT object has been garbage collected (and even the Python process has ended), then it is necessary to add `keep_intermediate=True`. I will add this nuance to the debugging tips.